contracts/MechMarketplace.sol

```diff
@@ -0,0 +1,438 @@
+        bytes memory data,
+        address priorityMech,
+        uint256 responseTimeout
+    ) external payable returns (uint256 requestId, uint256 requestIdWithNonce) {
+        // Reentrancy guard
+        if (_locked > 1) {
+            revert ReentrancyGuard();
+        }
+        _locked = 2;
+
+        // Check for zero address
+        if (priorityMech == address(0)) {
+            revert ZeroAddress();
+        }
+        // Agent mech itself cannot post a request
+        if (mapMechRegistrations[msg.sender]) {
+            revert UnauthorizedAccount(msg.sender);
+        }
+        // Check that priority mech is registered
+        if (!mapMechRegistrations[priorityMech]) {
+            revert UnauthorizedAccount(priorityMech);
+        }
+        // responseTimeout bounds
+        if (responseTimeout < minResponseTimeout || responseTimeout > maxResponseTimeout) {
+            revert OutOfBounds(responseTimeout, minResponseTimeout, maxResponseTimeout);
+        }
+        // responseTimeout limits
+        if (responseTimeout + block.timestamp > type(uint32).max) {
+            revert Overflow(responseTimeout + block.timestamp, type(uint32).max);
+        }
+        // Check for non-zero data
+        if (data.length == 0) {
+            revert ZeroValue();
+        }
+
+        // Get the request Id
+        requestId = getRequestId(msg.sender, data);
+        requestIdWithNonce = getRequestIdWithNonce(msg.sender, data, mapNonces[msg.sender]);
+
+        // Update sender's nonce
+        mapNonces[msg.sender]++;
+
+        // Get mech delivery info struct
+        MechDelivery storage mechDelivery = mapRequestIdDeliveries[requestIdWithNonce];
+
+        // Record priorityMech and response timeout
+        mechDelivery.priorityMech = priorityMech;
+        // responseTimeout from relative time to absolute time
+        mechDelivery.responseTimeout = uint32(responseTimeout + block.timestamp);
+        // Record request account
+        mechDelivery.account = msg.sender;
+
+        // Increase mech requester karma
+        mapRequesterMechKarma[msg.sender][priorityMech]++;
+
+        // Increase the number of undelivered requests
+        numUndeliveredRequests++;
+        // Increase the total number of requests
+        numTotalRequests++;
+
+        // Process request by a specified priority mech
+        IMech(priorityMech).request{value: msg.value}(msg.sender, data, requestId, requestIdWithNonce);
+
+        emit MarketplaceRequest(msg.sender, priorityMech, requestId, requestIdWithNonce, data);
+
+        _locked = 1;
+    }
+
+    /// @dev Delivers a request.
+    /// @notice This function can only be called by the agent mech delivering the request.
+    /// @param requestId Request id.
+    /// @param requestIdWithNonce Request id with nonce.
+    /// @param requestData Self-descriptive opaque data-blob.
+    function deliver(uint256 requestId, uint256 requestIdWithNonce, bytes memory requestData) external {
```

Comment:
Symmetrically to request going through marketplace to the mech, mech marketplace delivers it when triggered by the mech delivery.